integrated_channels/canvas/client.py

```diff
@@ -44,7 +47,78 @@ def update_content_metadata(self, serialized_data):
     def delete_content_metadata(self, serialized_data):
         pass
 
-    def _get_oauth_access_token(self, client_id, client_secret):
+    def _post(self, url, data, scope):
+        """
+        Make a POST request using the session object to a Canvas endpoint.
+
+        Args:
+            url (str): The url to send a POST request to.
+            data (str): The json encoded payload to POST.
+        """
+        self._create_session()
+        response = self.session.post(url, data=data)
+        return response.status_code, response.text
+
+    def _delete(self, url, data, scope):
         """
-        TODO: get oauth token for canvas api access
+        Make a DELETE request using the session object to a Canvas endpoint.
+
+        Args:
+            url (str): The url to send a DELETE request to.
+            data (str): The json encoded payload to DELETE.
+        """
+        self._create_session()
+        response = self.session.delete(url, data=data)
+        return response.status_code, response.text
+
+    def _create_session(self):
         """
+        Instantiate a new session object for use in connecting with Canvas. Each enterprise customer
+        connecting to Canvas should have a single client session.
+        """
+        if self.session is None:
+            # Create a new session with a valid token
+            if self.session:
+                self.session.close()
+            oauth_access_token = self._get_oauth_access_token(
+                self.enterprise_configuration.client_id,
+                self.enterprise_configuration.secret,
+            )
+            session = requests.Session()
+            session.headers['Authorization'] = 'Bearer {}'.format(oauth_access_token)
+            session.headers['content-type'] = 'application/json'
+            self.session = session
+
+    def _get_oauth_access_token(self, client_id, client_secret):
+        """Uses the client id, secret and refresh token to request the user's auth token from Canvas.
+
+        Args:
+            client_id (str): API client ID
+            client_secret (str): API client secret
+
+        Returns:
+            access_token (str): the OAuth access token to access the Canvas API as the user
+        Raises:
+            HTTPError: If we received a failure response code from Canvas.
+            RequestException: If an unexpected response format was received that we could not parse.
+        """
+        auth_token_url = urljoin(
+            self.enterprise_configuration.canvas_base_url,
+            self.config.oauth_token_auth_path,
+        )
+        auth_token_params = {
+            'grant_type': 'refresh_token',
+            'client_id': client_id,
+            'client_secret': client_secret,
+            'state': self.enterprise_configuration.enterprise_customer.uuid,
+            'refresh_token': self.enterprise_configuration.refresh_token,
```

Comment:
so how do you get the first `refresh_token` ? Is that via the oauth/redirect button stuff?

Reply:
At the point where we're creating a client session to update/post/delete a course to canvas, the user should have already gone through the redirect/get code process. The auth code is a single use so we can't reliably use it to generate access tokens, so we need to have the refresh token from a new `oauth/complete` endpoint.

Reply:
hmm, will be worth checking where what is the expiry timeframe of that code? if it expires this code needs to run within that timeframe right? if nothing, we can/should document that somewhere.

Also, if the code expires how are we going to inform the user reading the logs of this? We may need a very specific message/exception for that case?